dial.go

```diff
@@ -284,30 +294,132 @@ func (c *fdConn) RemoteAddr() net.Addr {
+func (d fdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	file := os.NewFile(d.fd, "")
 	c, err := net.FileConn(file)
 	if err != nil {
 		return nil, fmt.Errorf("failed to dial: %w", err)
 	}
 
 	conn := new(tntConn)
-	conn.net = &fdConn{Conn: c, Addr: fdAddr{Fd: d.Fd}}
+	conn.net = &fdConn{Conn: c, Addr: fdAddr{Fd: d.fd}}
 
 	dc := &deadlineIO{to: opts.IoTimeout, c: conn.net}
 	conn.reader = bufio.NewReaderSize(dc, bufSize)
 	conn.writer = bufio.NewWriterSize(dc, bufSize)
 
-	_, err = rawDial(conn, d.RequiredProtocolInfo)
+	return conn, nil
+}
+
+// Dial makes FdDialer satisfy the Dialer interface.
+func (d FdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	dialer := ProtocolDialer{
+		Dialer: fdDialer{
+			fd: d.Fd,
+		},
+		RequiredProtocolInfo: d.RequiredProtocolInfo,
+	}
+
+	return dialer.Dial(ctx, opts)
+}
+
+// AuthDialer is a dialer-wrapper that does authentication of a user.
+type AuthDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// Authentication options.
+	Auth Auth
+	// Username is a name of a user for authentication.
+	Username string
+	// Password is a user password for authentication.
+	Password string
+}
+
+// Dial makes AuthDialer satisfy the Dialer interface.
+func (d AuthDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
 	if err != nil {
-		conn.net.Close()
-		return nil, err
+		return conn, err
+	}
+	greeting := conn.Greeting()
+	if greeting.Salt == "" {
+		conn.Close()
+		return nil, fmt.Errorf("failed to authenticate: " +
+			"an invalid connection without salt")
+	}
+
+	if d.Username == "" {
+		return conn, nil
 	}
 
+	protocolAuth := conn.ProtocolInfo().Auth
+	if d.Auth == AutoAuth {
+		if protocolAuth != AutoAuth {
+			d.Auth = protocolAuth
+		} else {
+			d.Auth = ChapSha1Auth
+		}
+	}
+
+	if err := authenticate(conn, d.Auth, d.Username, d.Password,
+		conn.Greeting().Salt); err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("failed to authenticate: %w", err)
+	}
 	return conn, nil
 }
 
+// ProtocolDialer is a dialer-wrapper that reads and fills the Greeting and
+// ProtocolInfo of a connection.
+type ProtocolDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// RequiredProtocol contains minimal protocol version and
+	// list of protocol features that should be supported by
+	// Tarantool server. By default, there are no restrictions.
+	RequiredProtocolInfo ProtocolInfo
+}
+
+// Dial makes ProtocolDialer satisfy the Dialer interface.
+func (d ProtocolDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
+	if err != nil {
+		return conn, err
+	}
+
+	greetingConn := greetingConn{
+		Conn: conn,
+	}
+	version, salt, err := readGreeting(greetingConn)
+	if err != nil {
+		greetingConn.Close()
+		return nil, fmt.Errorf("failed to read greeting: %w", err)
+	}
+	greetingConn.greeting = Greeting{
+		Version: version,
+		Salt:    salt,
+	}
+
```

Comment:
So, why is the ProtocolDialer makes greeting? Let's it make `GreetingDialer`.

Reply:
Updated.